plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadResponse;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.ExpirationStatus;
+import software.amazon.awssdk.services.s3.model.GetBucketLifecycleConfigurationRequest;
+import software.amazon.awssdk.services.s3.model.GetBucketLifecycleConfigurationResponse;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.LifecycleRuleFilter;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.S3Object;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+
+import javax.crypto.SecretKey;
+import javax.inject.Inject;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.nullToEmpty;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.trino.plugin.exchange.FileSystemExchangeManager.PATH_SEPARATOR;
+import static java.lang.Math.toIntExact;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static software.amazon.awssdk.core.client.config.SdkAdvancedClientOption.USER_AGENT_PREFIX;
+import static software.amazon.awssdk.core.client.config.SdkAdvancedClientOption.USER_AGENT_SUFFIX;
+
+public class S3FileSystemExchangeStorage
+        implements FileSystemExchangeStorage
+{
+    private static final String DIRECTORY_SUFFIX = "_$folder$";
+
+    private final Region region;
+    private final int multiUploadPartSize;
+    private final S3Client s3Client;
+    private final S3AsyncClient s3AsyncClient;
+
+    private final ExecutorService deleteExecutor = newCachedThreadPool(threadsNamed("s3-delete-%s"));
```

Comment:
Bound number of threads. Can be a constant.
If there are multiple places where executor is used it may make sense to inject same  instance everywhere via constructor.

Reply:
`S3FileSystemExchangeStorage` is a singleton.

Re "Bound number of threads": I just imitated `TrinoS3FileSystem`'s `uploadExecutor`, which didn't bound the number of threads either. When I look up the Trino codebase of `newCachedThreadPool`, nowhere bounded the number of threads.